src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -193,197 +331,93 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
             .build());
   }
 
-  private void launchDataNode(SchedulerDriver driver, Offer offer) {
-    launchNode(
-        driver,
-        offer,
-        HDFSConstants.DATA_NODE_ID,
-        Arrays.asList(HDFSConstants.DATA_NODE_ID),
-        HDFSConstants.NODE_EXECUTOR_ID);
-  }
-
-  private void launchInitialJournalNodes(SchedulerDriver driver, Collection<Offer> offers) {
-    for (int i = 0; i < conf.getJournalNodeCount(); i++) {
-      if (offers.size() > 0) {
-        Offer offer = offers.iterator().next();
-        pendingOffers.remove(offer.getId());
-        launchNode(
-            driver,
-            offer,
-            HDFSConstants.JOURNAL_NODE_ID,
-            Arrays.asList(HDFSConstants.JOURNAL_NODE_ID),
-            HDFSConstants.NODE_EXECUTOR_ID);
-      }
+  private boolean maybeLaunchJournalNode(SchedulerDriver driver, Offer offer) {
+    if (persistentState.journalNodeRunningOnSlave(offer.getHostname())) {
+      log.info(String.format("Already running journalnode on %s", offer.getHostname()));
+      return false;
+    } else {
+      launchNode(
+          driver,
+          offer,
+          HDFSConstants.JOURNAL_NODE_ID,
+          Arrays.asList(HDFSConstants.JOURNAL_NODE_ID),
+          HDFSConstants.NODE_EXECUTOR_ID);
+      return true;
     }
   }
 
-  private void launchInitialNameNodes(SchedulerDriver driver, Collection<Offer> offers) {
-    for (int i = 0; i < HDFSConstants.TOTAL_NAME_NODES; i++) {
-      if (offers.size() > 0) {
-        Offer offer = offers.iterator().next();
-        pendingOffers.remove(offer.getId());
-        launchNode(
-            driver,
-            offer,
-            HDFSConstants.NAME_NODE_ID,
-            Arrays.asList(HDFSConstants.NAME_NODE_ID, HDFSConstants.ZKFC_NODE_ID,
-                HDFSConstants.JOURNAL_NODE_ID),
-            HDFSConstants.NAME_NODE_EXECUTOR_ID);
-      }
+  private boolean maybeLaunchNameNode(SchedulerDriver driver, Offer offer) {
+    if (persistentState.nameNodeRunningOnSlave(offer.getHostname())) {
```

Comment:
Reconcile branch fixes this: https://github.com/Banno/hdfs/blob/reconcileTasks/src/main/java/org/apache/mesos/hdfs/Scheduler.java#L373